code/controllers/subsystem/movement/movement_types.dm

```diff
@@ -0,0 +1,871 @@
+ * flags - Set of bitflags that effect move loop behavior in some way. Check _DEFINES/movement.dm
+ *
+**/
+/datum/controller/subsystem/move_manager/proc/jps_move(moving,
+	chasing,
+	delay,
+	timeout,
+	repath_delay,
+	max_path_length,
+	minimum_distance,
+	obj/item/card/id/id,
+	simulated_only,
+	turf/avoid,
+	skip_first,
+	subsystem,
+	priority,
+	flags,
+	datum/extra_info)
+	return add_to_loop(moving,
+		subsystem,
+		/datum/move_loop/has_target/jps,
+		priority,
+		flags,
+		extra_info,
+		delay,
+		timeout,
+		chasing,
+		repath_delay,
+		max_path_length,
+		minimum_distance,
+		id,
+		simulated_only,
+		avoid,
+		skip_first)
+
+/datum/move_loop/has_target/jps
+	///How often we're allowed to recalculate our path
+	var/repath_delay
+	///Max amount of steps to search
+	var/max_path_length
+	///Minimum distance to the target before path returns
+	var/minimum_distance
+	///An ID card representing what access we have and what doors we can open. Kill me
+	var/obj/item/card/id/id
+	///Whether we consider turfs without atmos simulation (AKA do we want to ignore space)
+	var/simulated_only
+	///A perticular turf to avoid
+	var/turf/avoid
+	///Should we skip the first step? This is the tile we're currently on, which breaks some things
+	var/skip_first
+	///A list for the path we're currently following
+	var/list/movement_path
+	///Cooldown for repathing, prevents spam
+	COOLDOWN_DECLARE(repath_cooldown)
+
+/datum/move_loop/has_target/jps/setup(delay, timeout, atom/chasing, repath_delay, max_path_length, minimum_distance, obj/item/card/id/id, simulated_only, turf/avoid, skip_first)
+	. = ..()
+	if(!.)
+		return
+	src.repath_delay = repath_delay
+	src.max_path_length = max_path_length
+	src.minimum_distance = minimum_distance
+	src.id = id
+	src.simulated_only = simulated_only
+	src.avoid = avoid
+	src.skip_first = skip_first
+	if(istype(id, /obj/item/card/id))
+		RegisterSignal(id, COMSIG_PARENT_QDELETING, .proc/handle_no_id) //I prefer erroring to harddels. If this breaks anything consider making id info into a datum or something
+
+/datum/move_loop/has_target/jps/start_loop()
+	. = ..()
+	INVOKE_ASYNC(src, .proc/recalculate_path)
+
+/datum/move_loop/has_target/jps/Destroy()
+	id = null //Kill me
+	avoid = null
+	return ..()
+
+/datum/move_loop/has_target/jps/proc/handle_no_id()
+	SIGNAL_HANDLER
+	id = null
+
+//Returns FALSE if the recalculation failed, TRUE otherwise
+/datum/move_loop/has_target/jps/proc/recalculate_path()
+	if(!COOLDOWN_FINISHED(src, repath_cooldown))
+		return
+	COOLDOWN_START(src, repath_cooldown, repath_delay)
+	SEND_SIGNAL(src, COMSIG_MOVELOOP_JPS_REPATH)
+	movement_path = get_path_to(moving, target, max_path_length, minimum_distance, id, simulated_only, avoid, skip_first)
+
+/datum/move_loop/has_target/jps/move()
+	var/atom/movable/atom = moving
+	if(!length(movement_path))
+		INVOKE_ASYNC(src, .proc/recalculate_path)
+		if(!length(movement_path))
+			return FALSE
```

Comment:
The problem with the procs finishing in the wrong order and thus having the wrong path in there is valid even trough unlikely i will think a way up to check for the newest path even with the fallback solution.

Reply:
ok this beeing async really gives me a lot of headaches

Reply:
Ok so this can't be solved easily because of how fast the movement loop gets called again before it can finish recalculating it will again call the recal proc i think i will implement something that won't allow 2 recalculate procs on the same datum to be called at the same time that should do the trick